### ISSUE TITLE
Proposal: Remove or Improve Support for 24.3

### ISSUE THREAD

**Issue by justbur** [syl20bnr/spacemacs#3116]
Several reasons come to mind but the main one is that many of the layer packages require 24.4 including magit for example. 

The only common use of 24.3 that I know of is those using Ubuntu 14.04 LTS. For those people I suggest we post a correct and detailed set of build instructions for 24.5. [this](http://ubuntuhandbook.org/index.php/2014/10/emacs-24-4-released-install-in-ubuntu-14-04/) seems like a good start

**Comment by robbyoconnor** [comment]
:+1: I think I've said (bluntly) that this is something that people should know how to do...

**Comment by syl20bnr** [comment]
I'm all for it now :panda_face:

**Comment by person808** [comment]
Yes! I would love to use `with-eval-after-load` instead of `eval-after-load`

**Comment by stormpat** [comment]
:+1:

**Comment by CestDiego** [comment]
:+1:

**Comment by bmag** [comment]
There are definitely good reasons to drop support for 24.3 (`advice-add` for example), but I want to raise a possible issue. Some people work in companies that restrict installation of software, and if those people work on Ubuntu, then 24.3 may be the newest version of Emacs that they can use (assuming the company enforces the use of 14.04 LTS). This is only a problem if there is a large amount of spacemacs users that work in such circumstances. Unfortunately, I don't have any statistics to determine is such users actually exist. Does anybody here have relevant statistics? Are there any Emacs 24.3 users reading this? 

Other than that, I totally agree that we should drop 24.3. Emacs 24.4 was a significant improvement over 24.3, especially from a package developer's point of view, and I'd recommend any 24.3 user to upgrade his/her emacs. I am myself an Ubuntu (and Lubuntu) user, and it was really easy to upgrade to 24.5 - in fact I might have followed the same link that @justbur posted.

@robbyoconnor I agree that it is something people should know how to do, but it takes time to learn all that a person should know, and we should make it easy for those who don't know yet how to do so. Emacs 24.4 was actually the first serious program that I compiled from source, and that was how I learned.

**Comment by robbyoconnor** [comment]
@bmag -- they can compile it and run it from their home directory :)

**Comment by bmag** [comment]
Yes, hopefully :) I'd like to hear the opinion of someone who actually has this problem, though

**Comment by cpaulik** [comment]
@robbyoconnor partly true. But imagine you do not have the build dependencies installed and would have to build them by hand. Not that I think that anybody has to work in an environment where this is the case.

**Comment by justbur** [comment]
@bmag I appreciate the issue you point out, but part of my reasoning here is that 24.3 support is not actually being respected at the moment. There are many layers that use packages requiring 24.4 and if we were doing a good job of supporting 24.3 we would have at least a warning or a workaround for those users and we don't, so they get failures that look like bugs. 

If we continue with 24.3, it really should be made more obvious to users what they can and can't use with that version. To be nice at least.

**Comment by justbur** [comment]
On the compiling issue, I think most people can do it but until you understand all the dependencies and the build tools for your system it can be error prone. 

You also have to imagine someone who heard about this Spacemacs thing and wants to try it. Saying "you have to compile emacs to do that go figure out how" is plenty of reason not to. Sorry but nobody needs Spacemacs that bad.

**Comment by TheBB** [comment]
Just for the record, I don't feel like "just compile Emacs" is a suitable answer to give users who can't, for whatever reason, install a recent version the normal way. The latest Ubuntu LTS is not an obscure operating system, and Spacemacs has (it appears to me) a certain ambition in user-friendliness that is incompatible with that approach.

As far as I can tell, the packages that are incompatible with 24.3 are all in optional layers. (This was the case a while ago, at least.) It would be a different proposition if either of the spacemacs layers included an incompatible package. It's not in my opinion out of the question that specific layers should have version requirements that are stricter than the spacemacs layers.

That said, magit is so ubiquitous that it might as well be part of the core layers so let's not do that in this case. Supporting the old version of magit in parallel is presumably a nightmare. So I'm fine with upping the version requirements *in this case* since magit is forcing our hand in a sense, but *not* because it's easy to compile Emacs. It may be, but I think people in general shouldn't have to.

**Comment by cpaulik** [comment]
Well in case of magit there is the additional problem that you need a git version that is newer than the 14.04LTS one.

**Comment by syl20bnr** [comment]
I have to think about it, it is easy to support magit for 24.3 like it is already done for paradox. To support it we conditionally switch to a local package in 24.3 or we may be able to use a recipe.

**Comment by syl20bnr** [comment]
OMG I love Spacemacs, let's do a 24.3 compatibility layer that users can add at the end of their layer list.

Let's don't use the new add-functions functions for now or make a Spacemacs macro to dispatch to the right functions.

**Comment by justbur** [comment]
Actually the magit issue may be easily resolved soon

https://github.com/magit/magit/issues/2277
https://github.com/magit/magit/issues/2269

**Comment by justbur** [comment]
It would be nice if each layer declared its minimum required emacs version and the configuration-layer code checked that minimum and issued a helpful error message before downloading the packages. 

I think that should be done regardless of what happens here.

**Comment by TheBB** [comment]
Yep, :+1: to that.

**Comment by TheBB** [comment]
They should get an error message in the init file too... now you just silently get stock Emacs.

**Comment by justbur** [comment]
To collect relevant information here, I've seen a few pointers to this [PPA](https://launchpad.net/~andreas-h/+archive/ubuntu/emacs) for getting Emacs 24.5 in Ubunut

**Comment by justbur** [comment]
@person808 Why don't you open a PR to backport `with-eval-after-load`? That one is really easy. The advice stuff is probably not so simple.

**Comment by syl20bnr** [comment]
The steps collected so far:
- [ ] back port with-evil-after-load
- [ ] make a macro to dispatch advice funcs
- [ ] add a :emacs or something keyword to package list (default is 24.4, only packages needing previous versions should specify it)
- [ ] display minimum emacs version warnings for both init.el and packages
- [ ] create a 24.3 compatibility layer taking ownership of all :emacs < 24.4 packages and that must be placed at the me of user layer list

**Comment by justbur** [comment]
@person808 @syl20bnr #3127

**Comment by StreakyCobra** [comment]
I'm closing this issue as the discussion seems to be over and points have been reported in #3198.
